File: Ruby/lib/civitas-ruby/civitas_juego.rb
require_relative 'jugador'
require_relative 'tablero'
require_relative 'mazo_sorpresas'
require_relative 'sorpresa'
require_relative 'tipo_sorpresa'
require_relative 'operaciones_juego'
require_relative 'gestor_estados'

module Civitas
  class CivitasJuego
    
    def initialize(nombres)
      @indice_jugador_actual = Dado.instance.quien_empieza(nombres.size)
      @gestor_estados = Gestor_estados.new
      @estado = @gestor_estados.estado_inicial
      @tablero
      @mazo = MazoSorpresas.new
      @jugadores = []
      nombres.each { |nombre| @jugadores << Jugador.new(nombre)}
      inicializar_tablero(@tablero)
      inicializar_mazo_sorpresas(@mazo)
    end

    def cancelar_hipoteca(ip)
      jugador_actual.cancelar_hipoteca(ip)
    end

    def comprar
      jugador_a = jugador_actual
      num_casilla_a = jugador_a.num_casilla_actual
      casilla = @tablero.casillas[num_casilla_a]
      titulo = casilla.titulo_propiedad
      res = jugador_a.comprar(titulo)
    end

    def construir_casa(ip)
      jugador_a = jugador_actual
      jugador_a.construir_casa(ip)
    end

    def construir_hotel(ip)
      jugador_a = jugador_actual
      jugador_a.construir_hotel(ip)
    end

    def final_del_juego
      fin = false
      i = 0
      @jugadores.each { |i| fin = true if i.en_bancarrota}
      fin
    end

    def casilla_actual
      jugador_a = jugador_actual
      @tablero.casillas[jugador_a.num_casilla_actual]
    end

    def jugador_actual
      @jugadores[@indice_jugador_actual]
    end

    def hipotecar(ip)
      jugador_actual.hipotecar(ip)
    end

    def info_jugador_texto
      jugador_actual.to_s
    end

    def salir_carcel_pagando
      jugador_actual.salir_carcel_pagando
    end

    def salir_carcel_tirando
      jugador_actual.salir_carcel_tirando
    end

    def siguiente_paso
      jugador_a = jugador_actual
      operacion = @gestor_estados.operaciones_permitidas(jugador_a, @estado)
      case operacion
      when OperacionesJuego::PASAR_TURNO
        pasar_turno
        @estado = siguiente_paso_completado(operacion)
      when OperacionesJuego::AVANZAR
        avanza_jugador
        @estado = siguiente_paso_completado(operacion)
      end
      operacion
    end

    def siguiente_paso_completado(operacion)
      @estado = @gestor_estados.siguiente_estado(jugador_actual, @estado, operacion)
    end

    def vender(ip)
      jugador_actual.vender(ip)
    end

    private

    def avanza_jugador
      jugador = jugador_actual
      posicion_actual = jugador.num_casilla_actual
      tirada = Dado.instance.tirar
      posicion_nueva = @tablero.nueva_posicion(posicion_actual, tirada)
      casilla = @tablero.casillas[posicion_nueva]
      contabilizar_pasos_por_salida(jugador)
      jugador.mover_a_casilla(posicion_nueva)
      casilla.recibe_jugador(@indice_jugador_actual, @jugadores)
      contabilizar_pasos_por_salida(jugador)
    end

    def contabilizar_pasos_por_salida(jugador_a)
      while @tablero.por_salida > 0
        jugador_a.pasa_por_salida
      end
    end

    def inicializar_mazo_sorpresas(tablero)
      @mazo.al_mazo(Sorpresa.new_evita_carcel(TipoSorpresa::SALIRCARCEL, @mazo))
      @mazo.al_mazo(Sorpresa.new_envia_carcel(TipoSorpresa::IRCARCEL, @tablero))
      @mazo.al_mazo(Sorpresa.new_envia_casilla(TipoSorpresa::IRCASILLA, @tablero, 0, "\"Envía a salida\""))
      @mazo.al_mazo(Sorpresa.new_envia_casilla(TipoSorpresa::IRCASILLA, @tablero, 10, "\"Envía al descanso\""))
      @mazo.al_mazo(Sorpresa.new_sorpresa(TipoSorpresa::PAGARCOBRAR, -13370, "\"Pierde 13370\""))
      @mazo.al_mazo(Sorpresa.new_sorpresa(TipoSorpresa::PAGARCOBRAR, 3000, "\"Recibe 3000\""))
      @mazo.al_mazo(Sorpresa.new_sorpresa(TipoSorpresa::PORCASAHOTEL, -700, "\"Pierde 700 por cada edificación\""))
      @mazo.al_mazo(Sorpresa.new_sorpresa(TipoSorpresa::PORCASAHOTEL, 420, "\"Recibe 420 por cada edificación\""))
      @mazo.al_mazo(Sorpresa.new_sorpresa(TipoSorpresa::PORJUGADOR, -1000, "\"Paga 1000 a cada jugador\""))
      @mazo.al_mazo(Sorpresa.new_sorpresa(TipoSorpresa::PORJUGADOR, 900, "\"Cada jugador le paga 900\""))
    end

    def inicializar_tablero(mazo)
      @tablero = Tablero.new(5)
      #@tablero.añade_juez
      @tablero.añade_casilla(Casilla.new_calle(TituloPropiedad.new("A", 10, 10, 10, 10, 10)))
      @tablero.añade_casilla(Casilla.new_calle(TituloPropiedad.new("B", 10, 10, 10, 10, 10)))
      @tablero.añade_casilla(Casilla.new_calle(TituloPropiedad.new("C", 10, 10, 10, 10, 10)))
      @tablero.añade_casilla(Casilla.new_sorpresa(@mazo, "S1"))
      @tablero.añade_casilla(Casilla.new_calle(TituloPropiedad.new("D", 10, 10, 10, 10, 10)))
      @tablero.añade_casilla(Casilla.new_calle(TituloPropiedad.new("E", 10, 10, 10, 10, 10)))
      @tablero.añade_casilla(Casilla.new_calle(TituloPropiedad.new("F", 10, 10, 10, 10, 10)))
      @tablero.añade_casilla(Casilla.new_calle(TituloPropiedad.new("G", 10, 10, 10, 10, 10)))
      @tablero.añade_casilla(Casilla.new_descanso("Descanso"))
      @tablero.añade_casilla(Casilla.new_calle(TituloPropiedad.new("H", 10, 10, 10, 10, 10)))
      @tablero.añade_casilla(Casilla.new_calle(TituloPropiedad.new("I", 10, 10, 10, 10, 10)))
      @tablero.añade_casilla(Casilla.new_sorpresa(@mazo, "S2"))
      @tablero.añade_casilla(Casilla.new_calle(TituloPropiedad.new("J", 10, 10, 10, 10, 10)))
      @tablero.añade_juez
      @tablero.añade_casilla(Casilla.new_calle(TituloPropiedad.new("K", 10, 10, 10, 10, 10)))
      @tablero.añade_casilla(Casilla.new_calle(TituloPropiedad.new("L", 10, 10, 10, 10, 10)))
      @tablero.añade_casilla(Casilla.new_impuesto(2500, "Impuesto"))
      @tablero.añade_casilla(Casilla.new_sorpresa(@mazo, "S3"))
    end

    def pasar_turno
      @indice_jugador_actual += 1
      @indice_jugador_actual = @indice_jugador_actual % @jugadores.size
    end

    def ranking
      @jugadores.sort
    end

  end
end

File: Java/src/civitas/CivitasJuego.java
package civitas;

import java.util.ArrayList;
import java.util.Collections;

public class CivitasJuego {

  private int indiceJugadorActual;
  private MazoSorpresas mazo;
  private EstadosJuego estado;
  private GestorEstados gestorEstados;
  private ArrayList<Jugador> jugadores;
  private Tablero tablero;

  private void avanzaJugador() {
    Jugador jugadorActual = getJugadorActual();
    int posicionActual = jugadorActual.getNumCasillaActual();
    int tirada = Dado.getInstance().tirar();
    int posicionNueva = tablero.nuevaPosicion(posicionActual, tirada);
    Casilla casilla = tablero.getCasilla(posicionNueva);
    this.contabilizarPasosPorSalida(jugadorActual);
    jugadorActual.moverACasilla(posicionNueva);
    casilla.recibeJugador(indiceJugadorActual, jugadores);
    this.contabilizarPasosPorSalida(jugadorActual);
  }

  public Boolean cancelarHipoteca(int ip) {
    return getJugadorActual().cancelarHipoteca(ip);
  }

  public CivitasJuego(ArrayList<String> nombres) {
    jugadores = new ArrayList<Jugador>();
    for(String i : nombres) {
        jugadores.add(new Jugador(i));
    }
    gestorEstados = new GestorEstados();
    estado = gestorEstados.estadoInicial();
    indiceJugadorActual = Dado.getInstance().quienEmpieza(jugadores.size());
    mazo = new MazoSorpresas();
    inicializarTablero(mazo);
    inicializarMazoSorpresas(tablero);
  }

  public Boolean comprar() {
    Jugador jugadorActual = getJugadorActual();
    Casilla casilla = getCasillaActual();
    TituloPropiedad titulo = casilla.getTituloPropiedad();
    Boolean res = jugadorActual.comprar(titulo);
    return res;
  }

  public Boolean construirCasa(int ip) {
    return getJugadorActual().construirCasa(ip);
  }

  public Boolean construirHotel(int ip) {
    return getJugadorActual().construirHotel(ip);
  }

  private void contabilizarPasosPorSalida(Jugador jugadorActual) {
    while (tablero.getPorSalida() > 0) {
      jugadorActual.pasaPorSalida();
    }
  }

  public Boolean finalDelJuego() {
    for (Jugador i : jugadores) {
      if (i.enBancarrota()) {
        return true;
      }
    }
    return false;
  }

  public Casilla getCasillaActual() {
    return tablero.getCasilla(getJugadorActual().getNumCasillaActual());
  }

  public Jugador getJugadorActual() {
    return jugadores.get(indiceJugadorActual);
  }

  public Boolean hipotecar(int ip) {
    return getJugadorActual().hipotecar(ip);
  }

  public String infoJugadorTexto() {
    return getJugadorActual().toString();
  }

  private void inicializarMazoSorpresas(Tablero tablero) {
    mazo.alMazo(new Sorpresa(TipoSorpresa.IRCARCEL, tablero));
    mazo.alMazo(new Sorpresa(TipoSorpresa.SALIRCARCEL, mazo));
    mazo.alMazo(new Sorpresa(TipoSorpresa.PAGARCOBRAR, 1000, "Recibe 1000$"));
    mazo.alMazo(new Sorpresa(TipoSorpresa.PAGARCOBRAR, -1000, "Paga 1000$"));
    mazo.alMazo(new Sorpresa(TipoSorpresa.IRCASILLA, tablero, 10, "Va a Parking"));
    mazo.alMazo(new Sorpresa(TipoSorpresa.IRCASILLA, tablero, 0, "Va a Salida"));
    mazo.alMazo(new Sorpresa(TipoSorpresa.PORCASAHOTEL, tablero, -800, "Paga impuestos de 800$ por cada construcción"));
    mazo.alMazo(new Sorpresa(TipoSorpresa.PORCASAHOTEL, tablero, 300, "Recibe honorarios de 300$ por cada edificación"));
    mazo.alMazo(new Sorpresa(TipoSorpresa.PORJUGADOR, tablero, -100, "Todos los jugadores reciben 100$ de ti"));
    mazo.alMazo(new Sorpresa(TipoSorpresa.PORJUGADOR, tablero, 50, "Todos los jugadores te pagan 50$"));
  }

  private void inicializarTablero(MazoSorpresas mazo) {
    tablero = new Tablero(5);        
    tablero.añadeCasilla(new Casilla(new TituloPropiedad("A", 10, 10, 10, 10)));
    // tablero.añadeJuez();
    tablero.añadeCasilla(new Casilla(new TituloPropiedad("B", 10, 10, 10, 10)));
    tablero.añadeCasilla(new Casilla(new TituloPropiedad("C", 10, 10, 10, 10)));
    tablero.añadeCasilla(new Casilla(mazo, "S1"));
    tablero.añadeCasilla(new Casilla(new TituloPropiedad("D", 10, 10, 10, 10)));
    tablero.añadeCasilla(new Casilla(new TituloPropiedad("E", 10, 10, 10, 10)));
    tablero.añadeCasilla(new Casilla(new TituloPropiedad("F", 10, 10, 10, 10)));
    tablero.añadeCasilla(new Casilla(new TituloPropiedad("G", 10, 10, 10, 10)));
    tablero.añadeCasilla(new Casilla("Descanso"));
    tablero.añadeCasilla(new Casilla(new TituloPropiedad("H", 10, 10, 10, 10)));
    tablero.añadeCasilla(new Casilla(new TituloPropiedad("I", 10, 10, 10, 10)));
    tablero.añadeCasilla(new Casilla(mazo, "S2"));
    tablero.añadeCasilla(new Casilla(new TituloPropiedad("J", 10, 10, 10, 10)));
    tablero.añadeJuez();
    tablero.añadeCasilla(new Casilla(new TituloPropiedad("K", 10, 10, 10, 10)));
    tablero.añadeCasilla(new Casilla(new TituloPropiedad("L", 10, 10, 10, 10)));
    tablero.añadeCasilla(new Casilla(2500.0f, "Impuesto")); // Si no cree que es cárcel
    tablero.añadeCasilla(new Casilla(mazo, "S3"));
    //System.out.println("NUM CASILLA CARCEL = " + tablero.getCarcel());
  }

  private void pasarTurno() {
    indiceJugadorActual = (indiceJugadorActual+1)%jugadores.size();
  }

  public ArrayList<Jugador> ranking() { // Debería ser privado?
    Collections.sort(jugadores);
    return jugadores;
  }

  public Boolean salirCarcelPagando() {
    return getJugadorActual().salirCarcelPagando();
  }

  public Boolean salirCarcelTirando() {
    return getJugadorActual().salirCarcelTirando();
  }

  public OperacionesJuego siguientePaso() {
    Jugador jugadorActual = jugadores.get(indiceJugadorActual);
    OperacionesJuego operacion = gestorEstados.operacionesPermitidas(jugadorActual, estado);
    if (operacion == OperacionesJuego.PASAR_TURNO) {
      pasarTurno();
      siguientePasoCompletado(operacion);
    }
    else if (operacion == OperacionesJuego.AVANZAR) {
      avanzaJugador();
      siguientePasoCompletado(operacion);
    }
    return operacion;
  }

  public void siguientePasoCompletado(OperacionesJuego operacion) {
    estado = gestorEstados.siguienteEstado(getJugadorActual(), estado, operacion);
  }
  
  public Boolean vender(int ip) {
    return getJugadorActual().vender(ip);
  }

}
